keps/sig-cli/20190920-future-of-kubectl-cp.md

```diff
@@ -0,0 +1,158 @@
+- Offer users example `kubectl exec`/ shell commands to address advanced options of copying files. (There are a few already in --help)
+
+### Non-Goals
+
+For either of these, a separate proposal weighing the cost/benefit would be required.  These are out of scope of this proposal to simplify `kubectl cp`:
+- Rewrite `kubectl cp` to not use tar, by modifying CRI as outlined partially [here](https://github.com/kubernetes/kubernetes/issues/58512). 
+- Rewrite `kubectl cp` to be functional in scratch based containers through use of ephemeral containers as outlined [here](https://github.com/kubernetes/kubernetes/issues/58512#issuecomment-533640463)
+
+## Proposal
+
+- `kubectl cp` should provide simple function of copying single file or directory between local environments and pods.
+- Identify and document `kubectl exec` commands to address more advanced options for copying files.  
+- In the least, provide users attempting to use `kubectl cp + symlinks/etc` with output showing comparable `kubectl exec ... tar` cmds.
+- It is up for a decision in this proposal whether the community prefers to implement the `shelling out to tar from within kubectl` or leave as suggestions in error output. 
+- Barring decision of the above, only the user stories listed below should be supported by `kubectl cp`.  If additional user stories are added via shelling out to tar from kubectl, 
+  those will be outlined below.    
+
+### User Stories
+
+#### UserA wants to copy a file or a directory from a local host to a pod (or a container in a pod)
+
+```console
+ $ kubectl cp localdir somens/somepod:/poddir (-c acontainer)
+ $ kubectl cp localdir somens/somepod:/poddir (-c acontainer)
+ $ kubectl cp localdir/filea somens/somepod:/poddir (-c acontainer)
+```
+
+#### UserB wants to copy a file or a directory from a pod to a local host
+
+```console
+ $ kubectl cp somens/somepod:/poddir localdir
+ $ kubectl cp somens/somepod:/poddir/filea localdir/filea
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+Any scripts or automation that currently rely on advanced features of `kubectl cp` will be broken.
+To mitigate, detailed information about why the command now fails as well as example `kubectl exec` options will be output. 
+
+## Design Details
+
+### Test Plan
+
+### Graduation Criteria
+
+### Upgrade / Downgrade Strategy
+
+Any function that is removed as a result of a CVE fix or other will be documented clearly.  Any function that is removed will result in the user getting 
+information about why this subcommand/option is no longer supported and also will be given alternative `kubectl exec` commands as a replacement in the failed command output. 
+
+### Version Skew Strategy
+
+## Implementation History
+
+## Drawbacks
+
+Users relying on automation with `kubectl cp` may be inconvenienced if options and features are removed from the command.  The motivation of improving security is weighed 
```

Comment:
I think `inconvenienced` is pretty light. Users will be forced to rewrite continuous delivery scripts. Applications and installers that they have written will fail and need to be rewritten. It's possible that source for images that rely on this will be lost, etc etc.

This is a breaking change and we need to take any such change very seriously for the impact that it has on our users.

Reply:
I do agree the wording is not the best, I don't think we can do more than inform users :disappointed:

Reply:
updated wording